File: frontend/src/api/metrics.ts
import { api } from "@/api/client";
import type {
  DiscoveryChannelMetric,
  IndustryMetric,
  MetricsFilters,
  ObjectionMetric,
  OverviewResponse,
  SellerMetric,
  SellerOption,
} from "@/api/types";

function buildQuery(filters: MetricsFilters): string {
  const params = new URLSearchParams();
  if (filters.from) params.set("from", filters.from);
  if (filters.to) params.set("to", filters.to);
  if (filters.seller_id !== undefined) params.set("seller_id", String(filters.seller_id));
  if (filters.industry) params.set("industry", filters.industry);
  if (filters.closed !== undefined) params.set("closed", String(filters.closed));
  if (filters.uncategorized !== undefined) {
    params.set("uncategorized", String(filters.uncategorized));
  }
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

export const fetchOverview = (filters: MetricsFilters) =>
  api.get<OverviewResponse>(`/api/v1/metrics/overview${buildQuery(filters)}`);

export const fetchSellerMetrics = (filters: MetricsFilters) =>
  api.get<SellerMetric[]>(`/api/v1/metrics/by-seller${buildQuery(filters)}`);

export const fetchIndustryMetrics = (filters: MetricsFilters) =>
  api.get<IndustryMetric[]>(`/api/v1/metrics/by-industry${buildQuery(filters)}`);

export const fetchObjections = (filters: MetricsFilters, limit = 10) => {
  const params = new URLSearchParams(buildQuery(filters).slice(1));
  params.set("limit", String(limit));
  const qs = params.toString();
  return api.get<ObjectionMetric[]>(
    `/api/v1/metrics/objections${qs ? `?${qs}` : ""}`,
  );
};

export const fetchDiscoveryChannels = (filters: MetricsFilters) =>
  api.get<DiscoveryChannelMetric[]>(
    `/api/v1/metrics/discovery-channels${buildQuery(filters)}`,
  );

export const fetchSellers = () => api.get<SellerOption[]>("/api/v1/sellers");

File: backend/app/api/v1/metrics.py
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session
from app.models.enums import Industry
from app.schemas.metrics import (
    DiscoveryChannelMetric,
    IndustryMetric,
    ObjectionMetric,
    OverviewResponse,
    SellerMetric,
)
from app.services.metrics import (
    MetricsFilters,
    get_by_discovery_channel,
    get_by_industry,
    get_by_seller,
    get_overview,
    get_top_objections,
)

router = APIRouter(prefix="/metrics", tags=["metrics"])


def metrics_filters(
    from_date: Annotated[date | None, Query(alias="from")] = None,
    to_date: Annotated[date | None, Query(alias="to")] = None,
    seller_id: int | None = None,
    industry: Industry | None = None,
    closed: bool | None = None,
) -> MetricsFilters:
    return MetricsFilters(
        from_date=from_date,
        to_date=to_date,
        seller_id=seller_id,
        industry=industry,
        closed=closed,
    )


SessionDep = Annotated[AsyncSession, Depends(get_session)]
FiltersDep = Annotated[MetricsFilters, Depends(metrics_filters)]


@router.get("/overview")
async def overview_endpoint(session: SessionDep, filters: FiltersDep) -> OverviewResponse:
    return await get_overview(session, filters)


@router.get("/by-seller")
async def by_seller_endpoint(session: SessionDep, filters: FiltersDep) -> list[SellerMetric]:
    return await get_by_seller(session, filters)


@router.get("/by-industry")
async def by_industry_endpoint(session: SessionDep, filters: FiltersDep) -> list[IndustryMetric]:
    return await get_by_industry(session, filters)


@router.get("/objections")
async def objections_endpoint(
    session: SessionDep,
    filters: FiltersDep,
    limit: Annotated[int, Query(ge=1, le=50)] = 10,
) -> list[ObjectionMetric]:
    return await get_top_objections(session, filters, limit=limit)


@router.get("/discovery-channels")
async def discovery_channels_endpoint(
    session: SessionDep, filters: FiltersDep
) -> list[DiscoveryChannelMetric]:
    return await get_by_discovery_channel(session, filters)
